qemu-armv8a/fastboot: Add startup script (NxInit init.rc)

JianyuWang0623 opened the pull request:
## Summary
Currently, fastbootd does not start automatically.
Compared with NSH, the newly added component NxInit("NuttX Init") has been selected because it supports restarting the daemon process when it exits abnormally.

The script of NxInit (init.rc) launch the necessary services:
  1. console
  2. fastboot

Related changes:
- https://github.com/apache/nuttx-apps/pull/3192
- https://github.com/apache/nuttx/pull/17215 
### The impact of NxInit and NSH on ELF size
Typical scenario: High sensitivity to code size, with no requirement for a shell.
| Config Changes                     | ELF Size(bytes)  |
|:-----------------------------------|----------:|
| defconfig                          | 7,008,056 |
| `+ETC_ROMFS`                       | 7,066,048 |
| `+INIT`                            | 7,127,536 |
| `-TELENT`                          | 7,127,024 |
| `-NSH_LIBRARY` <br>`-SYSTEM_NSH`   | 6,509,392 |

For the ELF file, Init requires 61,488 bytes, while NSH and NSH_LIBRARY together require 617,632 bytes(without non-essential functions: 220,184 bytes). Switching from NSH to NxInit can save a significant amount of space. (updated)

Depends on https://github.com/apache/nuttx-apps/pull/3192
## Impact
- qemu-armv8a:fastboot

## Testing
Selftest
```
LD: nuttx
Memory region         Used Size  Region Size  %age Used
CP: nuttx.hex
CP: nuttx.bin
[    0.010000] parsing /etc/init.d/init.rc
[    0.010000] line   1: 'on init'
[    0.010000] new section (on)
[    0.010000] add action 0x4042a270(init) to manager 0x4042a090
[    0.010000] line   2: '    class_start core'
[    0.010000] add command 0x4042a2d0(class_start) to action 0x4042a270
[    0.010000]   argv[0] 'class_start'
[    0.010000]   argv[1] 'core'
[    0.010000] line   3: 'service console sh'
[    0.010000] new section (service)
[    0.010000] line   4: '    class core'
[    0.010000] apply option[0] 0x40290b68 for class
[    0.010000]   argv[0] 'class'
[    0.010000]   argv[1] 'core'
[    0.020000] line   5: '    restart_period 1000'
[    0.020000] apply option[2] 0x40290b48 for restart_period
[    0.020000]   argv[0] 'restart_period'
[    0.020000]   argv[1] '1000'
[    0.020000] line   6: 'service fastboot fastbootd'
[    0.020000] new section (service)
[    0.020000] line   7: '    class core'
[    0.020000] apply option[0] 0x40290b68 for class
[    0.020000]   argv[0] 'class'
[    0.020000]   argv[1] 'core'
[    0.020000] line   8: '    restart_period 7000'
[    0.020000] apply option[2] 0x40290b48 for restart_period
[    0.020000]   argv[0] 'restart_period'
[    0.020000]   argv[1] '7000'
[    0.020000] == Dump Actions ==
[    0.020000] action 0x4042a270
[    0.020000]   event trigger: 'init'
[    0.020000]   argv[0] 'class_start'
[    0.020000]   argv[1] 'core'
[    0.020000] == Dump Services ==
[    0.020000] Service 0x4042a370 name 'console' path 'sh'
[    0.020000]   pid: 0
[    0.020000]   arguments:
[    0.020000]       [0] 'service'
[    0.020000]       [1] 'console'
[    0.020000]       [2] 'sh'
[    0.020000]   classes:
[    0.020000]     'core'
[    0.020000]   restart_period: 1000
[    0.020000] Service 0x4042a490 name 'fastboot' path 'fastbootd'
[    0.020000]   pid: 0
[    0.020000]   arguments:
[    0.020000]       [0] 'service'
[    0.020000]       [1] 'fastboot'
[    0.020000]       [2] 'fastbootd'
[    0.020000]   classes:
[    0.020000]     'core'
[    0.020000]   restart_period: 7000
[    0.020000] add event [0] 'boot'
[    0.040000] add event [1] 'init'
[    0.040000] add event [2] 'netinit'
[    0.040000] remove event [0] 'boot'
[    0.040000] add ready 0x4042a2d0 class_start
[    0.040000] executing command 'class_start'
[    0.040000] starting service 'console' ...
[    0.040000] service 'console' flag 0x0 add 0x4
[    0.040000]   +flag 'running'
[    0.040000] service 'console' flag 0x4 add 0x8
[    0.040000]   -flag 'restarting'
[    0.040000] service 'console' flag 0x4 add 0x1
[    0.040000]   -flag 'disabled'
[    0.040000] started service 'console' pid 4
[    0.040000] starting service 'fastboot' ...
[    0.040000] service 'fastboot' flag 0x0 add 0x4
[    0.040000]   +flag 'running'
[    0.040000] service 'fastboot' flag 0x4 add 0x8
[    0.040000]   -flag 'restarting'
[    0.040000] service 'fastboot' flag 0x4 add 0x1
[    0.040000]   -flag 'disabled'
[    0.040000] started service 'fastboot' pid 5
[    0.050000] remove event [1] 'init'
[    0.050000] remove event [2] 'netinit'
nsh> ps
  PID  PPID GROUP PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK            STACK    USED FILLED COMMAND
    0     0     0   0 FIFO     Kthread   - Ready              0000000000000000 0008160 0000928  11.3%  Idle_Task
    1     0     0 192 RR       Kthread   - Waiting  Semaphore 0000000000000000 0008096 0001040  12.8%  hpwork 0x403f5540 0x403f55c0
    2     0     0 100 RR       Kthread   - Waiting  Semaphore 0000000000000000 0008096 0001040  12.8%  lpwork 0x403f5490 0x403f5510
    3     0     3 100 RR       Task      - Waiting  Semaphore 0000000000000000 0008128 0002112  25.9%  init_main
    4     3     4 100 RR       Task      - Running            0000000000000000 0008128 0003344  41.1%  sh
    5     3     5 100 RR       Task      - Waiting  Semaphore 0000000000000000 0008128 0002144  26.3%  fastbootd
nsh>
nsh>
```

anchao commented on boards/arm64/qemu/qemu-armv8a/configs/fastboot/defconfig:
```diff
@@ -31,17 +31,19 @@ CONFIG_DRIVERS_VIRTIO_NET=y
 CONFIG_DRIVERS_VIRTIO_RNG=y
 CONFIG_DRIVERS_VIRTIO_SERIAL=y
 CONFIG_DRIVERS_VIRTIO_SOUND=y
+CONFIG_ETC_ROMFS=y
```
add new config named fastboot_init?

anchao commented on boards/arm64/qemu/qemu-armv8a/src/CMakeLists.txt:
```diff
@@ -30,4 +30,8 @@ if(CONFIG_LIBC_FDT)
 endif()
 
+if(CONFIG_FS_ROMFS)
+  list(APPEND RCSRCS etc/init.d/init.rc)
```
Separate this commit with defconfig

anchao reviewed (changes requested):
Is there any advantage to using Android init? Besides the service look-aside, I don't see any other benefits. Why is Android necessarily the right path?

JianyuWang0623 commented:
> Just a feedback after dev@ mailing list discussion, thank you for understanding / considering @JianyuWang0623 :
> 
> 1. We clearly need to use "Android System Init" naming here and update all kconfig/global variables names to clearly distinguish from current NuttX Init, Android System Init, and  SystemV Init (there is sill a change it will show up here one day), maybe others. "system_system" or "system_init" or "init_main" is really confusing and not self-explanatory, does not even point to "android_system_init". I know this is a bit longer but will provide important context and clear separation between completely different functionalities / implementations.
> 2. We need better documentation for this "Android System Init" as well as "Fastboot". These functionalities are welcome, people want to use it, but users / developers ask questions like: Why do we need another init system? What should happen with the previous ones? What does this new init system solve? How this relates to Fastboot? When, why, and how can I use it? What are best use cases with some examples? Please consider approach like explaining and convincing someone who wants to use it but has zero knowledge about the solution.



> Is there any advantage to using Android init? Besides the service look-aside, I don't see any other benefits. Why is Android necessarily the right path?

@anchao As the data above , NSH, when used as an initialization program, may seem relatively bloated in scenarios where code size is a major concern. NSH and Init can coexist, allowing users to choose based on their own needs. Init addresses issues such as the excessive code size of NSH when used as an initialization program, and its lack of management (e.g., restarting) for daemons.
| Config Changes                     | ELF Size(bytes)  |
|:-----------------------------------|----------:|
| defconfig                          | 7,008,056 |
| `+ETC_ROMFS`                       | 7,066,048 |
| `+INIT`                            | 7,127,536 |
| `-TELENT`                          | 7,127,024 |
| `-NSH_LIBRARY` <br>`-SYSTEM_NSH`   | 6,509,392 |

For the **ELF** file, Init requires 61,488 bytes, while NSH and NSH_LIBRARY together require 617,632 bytes. Switching from NSH to Init can save a significant amount of space(556,144 bytes).

JianyuWang0623 commented:
> So you're also going to provide capabilities similar to Android's Zygote/SystemServer/Firmware Manager, right? So if an app using NSH wants to use these capabilities, it must use Android Init, right? Why this strange design? Is it difficult for NSH to implement these capabilities?

@anchao The management of daemons/services does not refer to complex functions like Zygote/SystemServer, but rather the status monitoring of daemons such as sh and telnetd. For example, monitoring their exit and then restarting them with a specified interval, configuring the signal sent when stopping the daemons, and determining whether the entire system needs to restart if the daemons encounter exceptions.

anchao commented:
> @anchao The management of daemons/services does not refer to complex functions like Zygote/SystemServer, but rather the status monitoring of daemons such as sh and telnetd. For example, monitoring their exit and then restarting them with a specified interval, configuring the signal sent when stopping the daemons, and determining whether the entire system needs to restart if the daemons encounter exceptions.

1. System monitoring capabilities can also be implemented in nsh, rather than forcing developers to use init if they want enable this feature.
2. Android's init syntax is only a subset of the shell, with additional syntax keywords to accommodate Android startup requirements. Historically, init syntax has consistently added keywords to improve compatibility with shell, so why not just use the shell? Furthermore, multi-line descriptions are a strange design. Similar to SELinux rule files, we can use semicolons or colon to separate rules, as many systems do.

anchao commented:
> It's more strange to couple the initialization into shell, do you see any other POSIX OS implement in this way? From the concept, init must be one instance and never exit, but shell may be lunched many times(telnet, ssh, adb shell...) or killed.
> 
> @JianyuWang0623 's work is the right direction to decouple the initialization from nsh from the architecture. But, it's also has other benefit too:
> 
> 1. In many simple case(bootloader, ota), we can enable init, but disable nsh
> 2. Implement the different init solution

You are talking about a software requirement, I question how to implement it. I don't agree with Android init approach, but I think system monitoring is a valuable capability.

JianyuWang0623 commented on boards/arm64/qemu/qemu-armv8a/configs/fastboot/defconfig:
```diff
@@ -31,17 +31,19 @@ CONFIG_DRIVERS_VIRTIO_NET=y
 CONFIG_HAVE_CXXINITIALIZE=y
 CONFIG_IDLETHREAD_STACKSIZE=8192
-CONFIG_INIT_ENTRYPOINT="nsh_main"
+CONFIG_INIT_ENTRYPOINT="init_main"
```
@cederom Thank you. This is my view: whether it is the current NxInit, or SystemV Init (which may be added later), or other versions, from a functional perspective, the default value of PROGNAME should be "init". For example, in some Linux distributions, even systemd uses the widely recognized name "init":
```bash
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"

$ ls /usr/sbin/init -l
lrwxrwxrwx 1 root root 20 Jun  4 22:17 /usr/sbin/init -> /lib/systemd/systemd
```
```
SYSTEMD(1)
NAME
       systemd, init - systemd system and service manager
```

xiaoxiang781216 commented:
let's close related pr. I don't want to discuss this problem anymore.

anchao commented:
> I think that you are keeping a misunderstanding: this init system does NOT go in the nuttx kernel AT ALL.
> 
> I would be very interested to understand why you believe such thing.
> 
> It is AN APP that you can choose to use as user space entry point, or not. It does NOT replace nsh, it comes before. nsh could be started by nxinit, if you wish to do so, but other apps could be started instead. Just like in a linux PC where the init program is systemd, not bash.
> 
> I agree that no discussion deserve closing a pr prematurely, but I think @xiaoxiang781216 was a bit fed up trying to explain something while you believe something else.

When you reply to the comments, you'd better understand what these rcS/rc.sysinit in the nuttx kernel are used for. If a new type of initialization script is added, they use different syntax, what negative impact will it have on the operating system?

<img width="897" height="699" alt="image" src="https://github.com/user-attachments/assets/f7879571-7d38-4f07-8f4a-07e471868046" />

anchao commented:
> I cant deny it. Maintaining is harder. You have some insight.
> 
> You might see my point of view changing while writing the previous message :)
> 
> In retrospect it is obvious that having a new scripted init system requires new sets of init scripts.
> 
> Do you agree that we can work this out a bit later in the board dirs? We should discuss it separately.

As I said I have no objection to nxinit, you don't need to ask me for my opinion, if everyone wants it to be merged, then just merge it, I will complete the optimization work of nsh if free
